### PR TITLE
ARROW-18437: [C++][Parquet] Fix encoder for DELTA_BINARY_PACKED when flushing more than once

### PR DESCRIPTION
**Pull request by mapleFU**
When flush more than one block, `DELTA_BINARY_PACKED` will be corrupt, because it didn't reset the context after flush.

I'll add some tests this weekend.

### CODE REVIEW

File: cpp/src/parquet/encoding.cc [hunk 1]
```diff
@@ -2138,8 +2138,6 @@ class DeltaBitPackEncoder : public EncoderImpl, virtual public TypedEncoder<DTyp
           "but it's " +
           std::to_string(values_per_block % mini_blocks_per_block));
     }
-    // Reserve enough space at the beginning of the buffer for largest possible header.
-    PARQUET_THROW_NOT_OK(sink_.Advance(kMaxPageHeaderWriterSize));
```
Comment (wgtmac) [inline cpp/src/parquet/encoding.cc:2142]: Please make sure it still works correctly if there are zero values (i.e. an empty block).

Comment (mapleFU) [reply cpp/src/parquet/encoding.cc:2142]:
```c++
template <typename DType>
void DeltaBitPackEncoder<DType>::FlushBlock() {
  if (values_current_block_ == 0) {
    return;
  }
```

Don't worry, I will add some tests this weekend...

Comment (mapleFU) [reply cpp/src/parquet/encoding.cc]: Oh, you're right, I make this wrong. Let me just make it simple

Comment (wgtmac) [reply cpp/src/parquet/encoding.cc]: I was concerned that there is no sufficient space reserved for the header if the block contains empty values. :)

**Comment by mapleFU** [comment]
@wjones127 @pitrou @rok Mind take a took? I don't want to modify testing for a lot, so a just add a `ROUND_TRIP_TIME` to make it build multiple times. /cc @wgtmac

**Review by rok — CHANGES_REQUESTED** [review]
Thanks for looking into this @mapleFU, I didn't realize encoders get reused like this.
I'm adding a couple of naming suggestions.

**Comment by mapleFU** [comment]
Resolved now, thanks @rok

**Comment by mapleFU** [comment]
@pitrou Mind take a look?

**Review by pitrou — APPROVED** [review]

**Comment by mapleFU** [comment]
Retrigger CI, seems the macos error is not caused by me

**Comment by mapleFU** [comment]
@pitrou can we merge this patch?

**Comment by pitrou** [comment]
@rok Can you give this a final look and merge if ok?

**Comment by rok** [comment]
Merged. Thanks for noticing and fixing this @mapleFU !

**Pull request merged**